src/ripple/app/ledger/impl/LedgerCleaner.cpp

```diff
@@ -96,20 +101,24 @@ class LedgerCleanerImp
     //
     //--------------------------------------------------------------------------
 
-    void onPrepare ()
+    void onPrepare () override
     {
     }
 
-    void onStart ()
+    void onStart () override
     {
-        startThread();
+        thread_ = std::thread {&LedgerCleanerImp::run, this};
     }
 
-    void onStop ()
+    void onStop () override
     {
-        m_journal.info << "Stopping";
-        signalThreadShouldExit();
-        notify();
+        JLOG (j_.info) << "Stopping";
+        {
+            std::lock_guard<std::mutex> lock (mutex_);
+            state_ = State::shouldExit;
+            wakeup_.notify_one();
```

Comment:
Thanks Howard.  In "C++ Concurrency In Action", listing 4.1 it shows the `notify_one()` call under the lock, like I did here.  I could imagine situations where the `notify_one()` is more reliably associated with the `state_` change if it's done under the lock.  But I suspect it doesn't really make much difference in this situation.

Scott, the reason it's not a deadlock is that `std::condition_variable::wait()` releases the lock most of the time.  that's why you have to pass it a `std::unique_lock` (which can be `unlock()`ed).  It's interesting to see the techniques.  This is the first time I've messed with a `std::condition_variable`.